catboost/catboost_pruning.py

```diff
@@ -0,0 +1,155 @@
+            If you set multiple datasets as ``eval_set``, the index number of ``eval_set`` must be
+            included in the valid_name, e.g., ``validation_0`` or ``validation_1``
+            when ``eval_set`` contains two datasets.
+    """
+
+    def __init__(
+        self, trial: optuna.trial.Trial, metric: str, valid_name: str = "validation"
+    ) -> None:
+        self._trial = trial
+        self._metric = metric
+        self._valid_name = valid_name
+        self._pruned = False
+        self._message = ""
+
+    def after_iteration(self, info: Any) -> bool:
+        """Run after each iteration.
+
+        Args:
+            info:
+                A simpleNamespace containing iteraion, validation_name, metric_name
+                and history of losses.
+                For example ``namespace(iteration=2, metrics= {
+                'learn': {'Logloss': [0.6, 0.5]},
+                'validation': {'Logloss': [0.7, 0.6], 'AUC': [0.8, 0.9]}
+                })``.
+
+        Returns:
+            A boolean value. If :obj:`True`, the trial should be pruned.
+            Otherwise, the trial should continue.
+        """
+        step = info.iteration - 1
+        if self._valid_name not in info.metrics:
+            raise ValueError(
+                'The entry associated with the validation name "{}" '
+                "is not found in the evaluation result list {}.".format(self._valid_name, info)
+            )
+        metrics = info.metrics[self._valid_name]
+        if self._metric not in metrics:
+            raise ValueError(
+                'The entry associated with the metric name "{}" '
+                "is not found in the evaluation result list {}.".format(self._metric, info)
+            )
+        scores = metrics[self._metric]
+        current_score = scores[-1]
+        self._trial.report(current_score, step=step)
+        if self._trial.should_prune():
+            self._message = "Trial was pruned at iteration {}.".format(step)
+            self._pruned = True
+            return False
+        return True
+
+    def check_pruned(self) -> None:
+        """Check whether pruend."""
+        if self._pruned:
+            raise optuna.TrialPruned(self._message)
+
+
+def objective(trial: optuna.Trial) -> float:
+    data, target = load_breast_cancer(return_X_y=True)
+    train_x, valid_x, train_y, valid_y = train_test_split(data, target, test_size=0.25)
+
+    param = {
+        "objective": trial.suggest_categorical("objective", ["Logloss", "CrossEntropy"]),
+        "colsample_bylevel": trial.suggest_float("colsample_bylevel", 0.01, 0.1, log=True),
+        "depth": trial.suggest_int("depth", 1, 12),
+        "boosting_type": trial.suggest_categorical("boosting_type", ["Ordered", "Plain"]),
+        "bootstrap_type": trial.suggest_categorical(
+            "bootstrap_type", ["Bayesian", "Bernoulli", "MVS"]
+        ),
+        "used_ram_limit": "3gb",
+        "eval_metric": "AUC",
```

Comment:
Sorry, one more thing. Could you optimize `Accuracy` rather than `AUC`?
```suggestion
        "eval_metric": "Accuracy",
```

Reply:
I left a comment on this before, so if my comment mislead you, I'm sorry.